### DISCUSSION TITLE
How to create multiple instances of a "widget" for repeated use?

### DISCUSSION THREAD

**Discussion by AZarbade** [emilk/egui#5007] [Q&A]
Hi everyone,
**TLDR**: I'm trying to plot incoming data from stdin, there are multiple streams of data. I want to create a widget/container that will show the selected stream separately, in its respective instance of the same widget/container.

**Context**:
_Here, graph is just an example representation of the data._
![image](https://github.com/user-attachments/assets/b9738c2b-cc08-45d2-86cc-c6171844aa92)

**Problem**: As shown in the above image, when a stream is selected, its respective representation appears. However, when I select two or more streams, they overwrite or appear on top of the previous plot.
This is obvious when we look at how I have structured the code:
```Rust
        egui::CentralPanel::default().show(ctx, |_| {
            let stream_count = self.payload.lock().unwrap().stream_count;
            for i in 0..stream_count {
                let stream_key = format!("Stream_{i}");
                if let Some(is_plotted) = self.plot_tracker.get_mut(&stream_key) {
                    if *is_plotted {
                        let data = self.payload.lock().unwrap().get_plotpoints(i);
                        let mut inner_applets = wrap_app::Showcase { stream_index: i };
                        inner_applets.show(ctx, data);
                    }
                }
            }
        });
```

**Question**: Is there any way to create instances for a widget/container that I could use to display different data separately from each other? I am aiming to have one container that, based on selection, displays data in different ways.
I am pretty new to this "application development" side of things, so pardon me if there are any obvious mistakes.

**Comment by AZarbade** [comment]
Figured it out.

Initially, I was only calling `egui::Window::new()` once for the entire loop. Calling that _within_ the loop solved the issue. Also, each new window has to have an unique ID.

```Rust
    pub fn render_plot(&mut self, ctx: &egui::Context) {
        for index in 0..stream_count {
            if let Some(is_plotted) = payload.plot_tracker.get(&stream_key) {
                if *is_plotted {
                    // stream_id is **important**. Unique ID for each stream
                    let stream_id = egui::Id::new(stream_key);

                    // here, call new window explicitly for each stream with respective ID
                    egui::Window::new("")
                        .id(stream_id)
                        .default_size(vec2(512., 256.))
                        .show(ctx, |ui| {
                            ui.ctx().request_repaint();
                            let data = payload.get_plotpoints(index);
                            let plot = Plot::new("plot");
                            plot.show(ui, |ui| {
                                ui.line(Line::new(data));
                            });
                        });
                }
            }
        }
    }
```
Then the main app looks like this,

```Rust
        egui::CentralPanel::default().show(ctx, |_| {
            self.render_plot(ctx);
        });
```

Final look,
![image](https://github.com/user-attachments/assets/989f9121-7e68-463c-a440-d2bc817f6565)
